### PR TITLE
feat(oidc-prompt): OIDC Prompt API & CLI

### PR DESCRIPTION
**Pull request by elimt**
Boundary OIDC method does not currently support passing in prompts during authentication. This change adds the capability of passing OIDC prompts. Prompts are optional OIDC parameters that determine the behaviour of the authentication server: https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest

# Changes

- Add `prompt` option for OIDC auth method CLI create and update
- Pass configured prompt during OIDC authentication
- Add `prompt` API validation for create and update
- Add missed tests for prompt, immutable fields test, handler test

## API
1. Create OIDC auth method with `select_account` prompt
```
curl --location 'http://127.0.0.1:9200/v1/auth-methods' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer $TOKEN' \
--data '{
    "type": "oidc",
    "name": "OIDC Name",
    "description": "OIDC Description",
    "scope_id": "global",
    "attributes": {
        "issuer": "https://sts.windows.net/TENANT_ID/",
        "client_id": "$CLIENT_ID",
        "client_secret": "$CLIENT_SECRET",
        "api_url_prefix": "http://localhost:9200",
        "disable_discovered_config_validation": false,
        "dry_run": false,
        "signing_algorithms": ["RS256"],
        "prompts": ["select_account"]
    }
}'
```
2. Update OIDC auth method with multiple prompts
```
curl --location --request PATCH 'http://127.0.0.1:9200/v1/auth-methods/{id}' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer $TOKEN' \
--data '{
    "version": 1,
    "attributes": {
        "prompts": ["select_account", "consent"]
    }
}'
```
## CLI
1. Create OIDC auth method with `select_account` prompt
```
boundary auth-methods create oidc \                                         
  -issuer "https://sts.windows.net/TENANT_ID/" \
  -client-id "$CLIENT_ID" \
  -client-secret "$CLIENT_SECRET" \
  -signing-algorithm RS256 \
  -api-url-prefix "$API_URL_PREFIX" \
  -name "azure" \
  -prompts "select_account"
  
Auth Method information:
  Created Time:         Thu, 16 Nov 2023 16:35:09 EST
  ID:                   amoidc_CFZkkKFTct
  Name:                 azure
  Type:                 oidc
  Updated Time:         Thu, 16 Nov 2023 16:35:09 EST
  Version:              1

  Scope:
    ID:                 global
    Name:               global
    Type:               global

  Authorized Actions:
    no-op
    read
    update
    delete
    change-state
    authenticate

  Authorized Actions on Auth Method's Collections:
    accounts:
      create
      list
    managed-groups:
      create
      list

  Attributes:
    api_url_prefix:     http://localhost:9200
    callback_url:       http://localhost:9200/v1/auth-methods/oidc:authenticate:callback
    client_id:         <CLIENT_ID>
    client_secret_hmac: <HMAC>
    issuer:             https://sts.windows.net/TENANT_ID/
    prompts:            [select_account]
    signing_algorithms: [RS256]
    state:              inactive  
```

2. Update OIDC auth method with multiple prompts
```
boundary auth-methods update oidc -id amoidc_CFZkkKFTct -prompts 'select_account' -prompt 'consent'

Auth Method information:
  Created Time:           Thu, 16 Nov 2023 16:35:09 EST
  ID:                     amoidc_CFZkkKFTct
  Is Primary For Scope:   true
  Name:                   azure
  Type:                   oidc
  Updated Time:           Thu, 16 Nov 2023 16:37:52 EST
  Version:                3

  Scope:
    ID:                   global
    Name:                 global
    Type:                 global

  Authorized Actions:
    no-op
    read
    update
    delete
    change-state
    authenticate

  Authorized Actions on Auth Method's Collections:
    accounts:
      create
      list
    managed-groups:
      create
      list

  Attributes:
    api_url_prefix:       http://localhost:9200
    callback_url:         http://localhost:9200/v1/auth-methods/oidc:authenticate:callback
    client_id:            <CLIENT_ID>
    client_secret_hmac:   <HMAC>
    issuer:               https://sts.windows.net/TENANT_ID/
    prompts:              [consent select_account]
    signing_algorithms:   [RS256]
    state:                active-public
```

### CODE REVIEW

File: internal/auth/oidc/service_start_auth.go [hunk 1]
```diff
@@ -118,6 +118,11 @@ func StartAuth(ctx context.Context, oidcRepoFn OidcRepoFactory, authMethodId str
 		oidcOpts = append(oidcOpts, oidc.WithScopes(am.ClaimsScopes...))
 	}
 
+	if len(am.Prompts) > 0 {
```
Comment (jimlambrt) [inline internal/auth/oidc/service_start_auth.go:121]: should we test this in `Test_StartAuth(...)`?

File: internal/auth/oidc/prompt.go [hunk 2]
```diff
@@ -31,6 +32,12 @@ var supportedPrompts = map[PromptParam]bool{
 	SelectAccount: true,
 }
 
+// SupportedPrompt returns true if the provided prompt is supported
+// by boundary.
+func SupportedPrompt(p PromptParam) bool {
```
Comment (jimlambrt) [inline internal/auth/oidc/prompt.go:37]: Sounds silly, but should we write a unit test for this?

File: internal/cmd/commands/authmethodscmd/oidc_funcs.go [hunk 3]
```diff
@@ -33,6 +33,7 @@ type extraOidcCmdVars struct {
 	flagAccountClaimMaps                  []string
 	flagDisableDiscoveredConfigValidation bool
 	flagDryRun                            bool
+	flagPrompts                           []string
```
Comment (jimlambrt) [inline internal/cmd/commands/authmethodscmd/oidc_funcs.go:36]: Just checking: did you test the CLI by hand and verify that it works as expected
with the new prompts?

File: internal/daemon/controller/handlers/authmethods/authmethod_service.go [hunk 4]
```diff
@@ -862,6 +862,7 @@ func toAuthMethodProto(ctx context.Context, in auth.AuthMethod, opt ...handlers.
 			AllowedAudiences:  i.GetAudClaims(),
 			ClaimsScopes:      i.GetClaimsScopes(),
 			AccountClaimMaps:  i.GetAccountClaimMaps(),
+			Prompts:           i.GetPrompts(),
```
Comment (jimlambrt) [inline internal/daemon/controller/handlers/authmethods/authmethod_service.go:865]: sort of a general question: should we add
prompts to the appropriate test in authmethod_service_test.go?

**Review by jimlambrt — CHANGES_REQUESTED** [review]
Looks great.  A few suggestions/questions.

File: internal/cmd/commands/authmethodscmd/oidc_funcs.go (hunk 3 above)
Comment (elimt) [reply internal/cmd/commands/authmethodscmd/oidc_funcs.go:36]: Yes I did. I'll do it and add the output to the PR

**Comment by elimt** [comment]
@jimlambrt I added some more tests and also updated the PR description to include the output of the prompt on Boundary CLI

**Review by jimlambrt — APPROVED** [review]
ty!
